File: rust/src/bin/p0067.rs
// Project Euler: Problem 67

euler::run_solver!(67);

static FILE_DATA: &str = include_str!("../../assets/0067_triangle.txt");

fn solve() -> String {
    compute(FILE_DATA).to_string()
}

fn compute(data: &str) -> u64 {
    parse_data(data)
        .into_iter()
        .rev()
        .reduce(|acc, e| {
            acc.windows(2)
                .map(|v| v[0].max(v[1]))
                .zip(e)
                .map(|(e1, e2)| e1 + e2)
                .collect()
        })
        .unwrap()[0]
}

fn parse_data(data: &str) -> Vec<Vec<u64>> {
    data.lines()
        .map(|s| {
            s.split_ascii_whitespace()
                .map(|elm| elm.parse::<u64>().unwrap())
                .collect()
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0067() {
        assert_eq!(compute(super::FILE_DATA), 7273);
    }
}

File: ocaml/src/p0067.ml
(* Project Euler: Problem 67 *)

open Core

let select_leaf lst = List.(map2_exn (tl_exn lst) (drop_last_exn lst) ~f:Int.max)

let calc_from_bottom l_lst =
  List.(
    fold_right
      (drop_last_exn l_lst)
      ~f:(fun a b -> map2_exn a (select_leaf b) ~f:( + ))
      ~init:(last_exn l_lst)
    |> hd_exn)
;;

let parse_data data =
  List.(map ~f:(Fun.compose (map ~f:Int.of_string) (String.split ~on:' ')) data)
;;

let compute str_lst = parse_data str_lst |> calc_from_bottom
let solve fname = compute (Euler.Task.read_file fname) |> Int.to_string

(* Test *)

let%test_unit "p067_triangle.txt" =
  [%test_eq: int] (compute (Euler.Task.read_file "./assets/p067_triangle.txt")) 7273
;;
